src/types/dict.rs

```diff
@@ -430,13 +430,12 @@ impl<'py> PyDictMethods<'py> for Bound<'py, PyDict> {
             key: Bound<'_, PyAny>,
         ) -> PyResult<Option<Bound<'py, PyAny>>> {
             let py = dict.py();
-            match unsafe {
-                ffi::PyDict_GetItemWithError(dict.as_ptr(), key.as_ptr())
-                    .assume_borrowed_or_opt(py)
-                    .map(Borrowed::to_owned)
-            } {
-                some @ Some(_) => Ok(some),
-                None => PyErr::take(py).map(Err).transpose(),
+            unsafe {
+                let mut result: *mut ffi::PyObject = std::ptr::null_mut();
+                match ffi::PyDict_GetItemRef(dict.as_ptr(), key.as_ptr(), &mut result) {
+                    std::os::raw::c_int::MIN..=0 => PyErr::take(py).map(Err).transpose(),
+                    1..=std::os::raw::c_int::MAX => Ok(result.assume_owned_or_opt(py)),
```

Comment:
Reading over after writing this, it occurs to me it might be more correct to define an enum that represents the possible return values from the function? Rather than allowing the possibility for it to return results besides -1, 0, and 1, which are the only values it's documented to return.

Reply:
Thanks for the pointer about `fetch`! I missed that in the docs. Much nicer.

Also splitting out the `0` case allowed me to catch a bug in my polyfill implementation, so thanks for making me catch that too!

Reply:
Great! 😂